### PR TITLE
Validate requested node provisioninging against available cr hardware

### PR DESCRIPTION
**Pull request by chrisdoherty4**
## Description
Add pre-flight validation to assert there are sufficient Tinkerbell CRs to satisfy the requested provisioning of control plane, etcd and worker nodes.

## Future work

There is a tight coupling between a few methods as the `ValidateHardwareConfig()` assumes responsibility for loading and parsing the hardware config that's required by `ValidateMinimumRequiredTinkerbellHardwareAvailable()`. This made it difficult to declare the `ValidateMinimumRequiredTinkerbellHardwareAvailable()` dependencies explicitly as parameters; we should really decoupled that to create much better testability.

### CODE REVIEW

File: pkg/providers/tinkerbell/validator.go [hunk 1]
```diff
@@ -5,12 +5,20 @@ import (
...
-	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
 	"github.com/aws/eks-anywhere/pkg/hardware"
 	"github.com/aws/eks-anywhere/pkg/logger"
 	"github.com/aws/eks-anywhere/pkg/networkutils"
 )
 
+// NodeMinimum defines the minimum number of nodes for the respective
+// node types.
+const (
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/validator.go:16]: Need to remove this, it was part of an additional validation I'd done.

File: pkg/providers/tinkerbell/tinkerbell.go [hunk 2]
```diff
@@ -179,6 +179,12 @@ func (p *tinkerbellProvider) SetupAndValidateCreateCluster(ctx context.Context,
 		return fmt.Errorf("failed setup and validations: %v", err)
 	}
 
+	// ValidateHardwareConfig performs a lazy load of hardware configuration. Given subsequent steps need the hardware
```
Comment (pokearu) [inline pkg/providers/tinkerbell/tinkerbell.go:182]: That's a good catch 😄  
Will do a follow up PR to decouple this.

Comment (chrisdoherty4) [reply pkg/providers/tinkerbell/tinkerbell.go:182]: I have some ideas around how you can do this that would make dependencies more explicit for each validation method if you'd like to chat.

Comment (pokearu) [reply pkg/providers/tinkerbell/tinkerbell.go:182]: Definitely! If you already started on it feel free to raise a PR with what you had in mind 👍

File: pkg/providers/tinkerbell/validator_helper_test.go [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package tinkerbell
+
+import "github.com/aws/eks-anywhere/pkg/hardware"
+
+// SetValidatorHardwareConfig sets v internal hardwareConfig member to cfg.
+func SetValidatorHardwareConfig(v *Validator, cfg hardware.HardwareConfig) {
+	v.hardwareConfig = cfg
+}
```
Comment (abhinavmpandey08) [inline pkg/providers/tinkerbell/validator_helper_test.go:8]: i would vote to just do this inline instead of having a separate function for this

**Review by abhinavmpandey08 — COMMENTED** [review]
awesome work! Just a few minor comments on convention and stuff

File: pkg/providers/tinkerbell/validator_helper_test.go (hunk 3 above)
Comment (chrisdoherty4) [reply pkg/providers/tinkerbell/validator_helper_test.go:8]: If you look at the packages they're different - I've used the `_test` package idiom for the unit tests. This is a helper function used in testing only, because it shouldn't be a part of the package API, for hacking a `Validator` instance removing the need to construct the `Validator` in full. It exists largely because of a tight coupling between reading hardware configuration and one of the validate methods that lazily loads the hardware config. As a rule of thumb you don't want io occurring in unit tests if you can help it. 

In short, this can't be done in-line using the `_test` idiom which is best practice for unit testing with some small exceptions.

File: pkg/providers/tinkerbell/validator_test.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	tinkv1alpha1 "github.com/tinkerbell/cluster-api-provider-tinkerbell/tink/api/v1alpha1"
+
+	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/hardware"
+	. "github.com/aws/eks-anywhere/pkg/providers/tinkerbell"
```
[the remark this answers is not in the document]
Comment (chrisdoherty4) [reply pkg/providers/tinkerbell/validator_test.go:11]: Sure, I'm indifferent. Its common to use dot notation in unit tests when following the `_test` package idiom but it doesn't really matter which way we do it.

**Review by abhinavmpandey08 — APPROVED** [review]
/lgtm

File: pkg/providers/tinkerbell/tinkerbell.go (hunk 2 above)
Comment (chrisdoherty4) [reply pkg/providers/tinkerbell/tinkerbell.go:182]: Working on it. I found some other oddities so I might propose something a little more elaborate.
